plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TemporalTimeTravelUtil.java

```diff
@@ -0,0 +1,292 @@
+    private static long searchTowardsTailParallel(TrinoFileSystem fileSystem, long entryNumber, String transactionLogDir, long epochMillis, Executor executor, int maxLinearSearchSize)
+            throws IOException
+    {
+        ImmutableList.Builder<Callable<Long>> versionSearchTasks = ImmutableList.builder();
+        versionSearchTasks.add(() -> searchTowardsTailLinear(fileSystem, entryNumber, entryNumber + maxLinearSearchSize - 1, transactionLogDir, epochMillis));
+
+        long start = entryNumber + maxLinearSearchSize;
+        while (fileSystem.newInputFile(getTransactionLogJsonEntryPath(transactionLogDir, start)).exists()) {
+            long head = start;
+            long tail = start + maxLinearSearchSize - 1;
+            versionSearchTasks.add(() -> searchTowardsTailLinear(fileSystem, head, tail, transactionLogDir, epochMillis));
+            start += maxLinearSearchSize;
+        }
+
+        try {
+            return processWithAdditionalThreads(versionSearchTasks.build(), executor).stream()
+                    .max(Long::compare)
+                    .orElse(-1L);
+        }
+        catch (ExecutionException e) {
+            throw new RuntimeException(e.getCause());
+        }
+    }
+
+    private static long searchTowardsTailLinear(TrinoFileSystem fileSystem, long start, long end, String transactionLogDir, long epochMillis)
+            throws IOException
+    {
+        long entryNumber = start;
+        long version = -1;
+        Optional<TransactionLogEntries> entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+        while (entryNumber <= end && entries.isPresent()) {
+            Optional<CommitInfoEntry> commitInfo = entries.get().getEntries(fileSystem)
+                    .map(DeltaLakeTransactionLogEntry::getCommitInfo)
+                    .filter(Objects::nonNull)
+                    .findFirst();
+
+            if (commitInfo.isEmpty()) {
+                entryNumber++;
+                entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+                continue;
+            }
+
+            if (commitInfo.map(TemporalTimeTravelUtil::getCommitInfoTimestamp).orElseThrow() > epochMillis) {
+                break;
+            }
+
+            version = entryNumber;
+
+            entryNumber++;
+            entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+        }
+
+        return version;
+    }
+
+    private static long searchTowardsHeadLinear(TrinoFileSystem fileSystem, long start, long end, String transactionLogDir, long epochMillis)
+            throws IOException
+    {
+        long entryNumber = end;
+        Optional<TransactionLogEntries> entries = getEntriesFromJson(entryNumber, transactionLogDir, fileSystem, DataSize.ofBytes(0));
+        while (start <= entryNumber && entries.isPresent()) {
```

Comment:
Consider the scenario (where the first few transaction log files are pruned) where `_delta_log` directory looks like :
- 2.json
- 3.json
- 3.checkpoint.parquet
- 4.json
- 5.json

If i understand correctly, consdiering that `start` is set to `0` and `0.json` is not present, then the algorithm returns `-1`  disregarding the checks  for `1.json` & `2.json` ....

Reply:
No. We are searching from end towards head here.